src/hotspot/os/linux/os_linux.cpp

```diff
@@ -194,15 +194,12 @@ julong os::Linux::available_memory() {
   julong avail_mem;
 
   if (OSContainer::is_containerized()) {
-    jlong mem_limit, mem_usage;
-    if ((mem_limit = OSContainer::memory_limit_in_bytes()) < 1) {
-      log_debug(os, container)("container memory limit %s: " JLONG_FORMAT ", using host value",
-                             mem_limit == OSCONTAINER_ERROR ? "failed" : "unlimited", mem_limit);
-    }
-    if (mem_limit > 0 && (mem_usage = OSContainer::memory_usage_in_bytes()) < 1) {
+    jlong mem_limit = OSContainer::memory_limit_in_bytes();
+    jlong mem_usage = OSContainer::memory_usage_in_bytes();
+    if (mem_limit > 0 && mem_usage < 1) {
```

Comment:
This will run into the `OSContainer::memory_usage_in_bytes()` is not cached problem (you should notice this that the logs are more noisy). A pattern like:

```
    jlong mem_limit = OSContainer::memory_limit_in_bytes();
    jlong mem_usage;
    if (mem_limit > 0 && (mem_usage = OSContainer::memory_usage_in_bytes()) < 1) {
    [...]
```

Works better due to short-circuiting `and`-expression.

Reply:
Good catch, sorry, I think it was that way in the first place (I've undone it over-zealously in the name of legibility). Fixed in ac6469310bc.